mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h

```diff
@@ -20,25 +20,49 @@
 namespace mlir {
 class Conv2dGenerator {
 public:
-  LogicalResult
-  parseConvDims(std::string &inputLayout, std::string &outputLayout,
-                std::string &filterLayout, int64_t groupSize, int64_t batchSize,
-                int64_t inputChannel, int64_t inputHeight, int64_t inputWidth,
-                int64_t outputChannel, int64_t outputHeight,
-                int64_t outputWidth, int64_t filterHeight, int64_t filterWidth,
-                SmallVector<int64_t, 5> &filterDimension,
-                SmallVector<int64_t, 5> &inputDimension,
-                SmallVector<int64_t, 5> &outputDimension);
+  struct Config {
+    std::string arch;
+    int num_cu;
+    bool xdlops;
+    std::string operation;
+    std::string dataTypeStr;
+    int dilationHeight, dilationWidth;
+    int strideHeight, strideWidth;
+    int paddingHeight, paddingWidth;
+    std::string filterLayout;
+    std::string inputLayout;
+    std::string outputLayout;
 
-  LogicalResult genConvModule(
-      std::string &arch, int num_cu, std::string &operation,
-      std::string &inputLayout, std::string &outputLayout,
-      std::string &filterLayout, const SmallVector<int64_t, 5> &filterDimension,
-      const SmallVector<int64_t, 5> &inputDimension,
-      const SmallVector<int64_t, 5> &outputDimension, int dilationHeight,
-      int dilationWidth, int strideHeight, int strideWidth, int paddingHeight,
-      int paddingWidth, ModuleOp &module, OpBuilder &builder,
-      std::string &kernelName, mlir::Type dataType, bool xdlops = false);
+    std::string kernelName;
+
+    int outputSize;
+    SmallVector<int64_t, 5> filterDimension;
+    SmallVector<int64_t, 5> inputDimension;
+    SmallVector<int64_t, 5> outputDimension;
+  };
+
+  Conv2dGenerator(const std::string &arch = "", int num_cu = 0,
+                  bool xdlops = false, const std::string &operation = "conv2d",
+                  const std::string &dataTypeStr = "f32",
+                  int dilationHeight = 1, int dilationWidth = 1,
+                  int strideHeight = 1, int strideWidth = 1,
+                  int paddingHeight = 0, int paddingWidth = 0,
+                  const std::string &filterLayout = "kcyx",
+                  const std::string &inputLayout = "nchw",
+                  const std::string &outputLayout = "nkhw",
+                  const std::string &kernelName = "");
+
+  const Config &getConfig() const { return config; }
```

Comment:
Could you think of a different way than returning the config as-is? I'm worried about how caller would use a partially constructed config object.